File: python/leetcode/spiral-matrix-iii.py
class Solution:
    def spiralMatrixIII(self, R, C, r0, c0):
        """
        :type R: int
        :type C: int
        :type r0: int
        :type c0: int
        :rtype: List[List[int]]
        """
        r = r0
        c = c0
        result = [[r, c]]
        visits = 1
        steps = 1
        while visits < R * C:
            # Go across to the right
            for i in range(steps):
                c += 1
                if r < R and r >= 0 and c < C and c >= 0:
                    result.append([r, c])
                    visits += 1

            # Go down
            for i in range(steps):
                r += 1
                if r < R and r >= 0 and c < C and c >= 0:
                    result.append([r, c])
                    visits += 1
            steps += 1

            # Go back to the left
            for i in range(steps):
                c -= 1
                if r < R and r >= 0 and c < C and c >= 0:
                    result.append([r, c])
                    visits += 1

            # Go up
            for i in range(steps):
                r -= 1
                if r < R and r >= 0 and c < C and c >= 0:
                    result.append([r, c])
                    visits += 1
            steps += 1
        return result

File: java/leetcode/spiral-matrix-iii.java
class Solution {
    public int[][] spiralMatrixIII(int R, int C, int r0, int c0) {
        int r = r0;
        int c = c0;
        int[][] result = new int[R * C][2];
        result[0][0] = r;
        result[0][1] = c;
        int visits = 1;
        int steps = 1;
        while (visits < R * C) {
            // across to the right
            for (int i = 0; i < steps; i++) {
                c++;
                if (r < R && r >= 0 && c < C && c >= 0) {
                    result[visits][0] = r;
                    result[visits][1] = c;
                    visits++;
                }
            }
        
            // down
            for (int i = 0; i < steps; i++) {
                r++;
                if (r < R && r >= 0 && c < C && c >= 0) {
                    result[visits][0] = r;
                    result[visits][1] = c;
                    visits++;
                }
            }
            steps++;
            
            // across to the right
            for (int i = 0; i < steps; i++) {
                c--;
                if (r < R && r >= 0 && c < C && c >= 0) {
                    result[visits][0] = r;
                    result[visits][1] = c;
                    visits++;
                }
            }
            
            // up
            for (int i = 0; i < steps; i++) {
                r--;
                if (r < R && r >= 0 && c < C && c >= 0) {
                    result[visits][0] = r;
                    result[visits][1] = c;
                    visits++;
                }
            }
            steps++;
        }
        return result;
    }
}